inference.py

```diff
@@ -43,112 +45,223 @@
+                window_array = _pad_diff(window_array, window_array.shape[0], window_array.shape[1], chunk_size)
+            window_array = _pad(window_array, chunk_size)
+
+            yield window_array, row, column
 
 
 @torch.no_grad()
-def segmentation(img_array,
+def segmentation(param,
                  input_image,
                  label_arr,
                  num_classes: int,
                  gpkg_name,
                  model,
                  chunk_size: int,
-                 num_bands: int,
                  device,
                  scale: List,
                  BGR_to_RGB: bool,
-                 debug=False):
-
-    # switch to evaluate mode
-    model.eval()
-    transforms = tta.Compose([tta.HorizontalFlip(), ])
-
-    WINDOW_SPLINE_2D = _window_2D(window_size=chunk_size, power=2.0)
-    WINDOW_SPLINE_2D = torch.as_tensor(np.moveaxis(WINDOW_SPLINE_2D, 2, 0), ).type(torch.float)
-    WINDOW_SPLINE_2D = WINDOW_SPLINE_2D.to(device)
-
-
-    metadata = add_metadata_from_raster_to_sample(img_array,
-                                                  input_image,
-                                                  meta_map=None,
-                                                  raster_info=None)
-
+                 tp_mem,
+                 debug=False,
+                 ):
     xmin, ymin, xmax, ymax = (input_image.bounds.left,
                               input_image.bounds.bottom,
                               input_image.bounds.right,
                               input_image.bounds.top)
-
     xres, yres = (abs(input_image.transform.a), abs(input_image.transform.e))
-    h, w, bands = img_array.shape
-    if num_bands < bands and debug:
-        logging.warning(f'Expected {bands} bands, for {num_bands}. Slicing off excess bands. Problems may occur')
-        img_array = img_array[:, :, :num_bands]
-    padding = int(round(chunk_size * (1 - 1.0 / 2.0)))
-    step = int(chunk_size / 2.0)
-    img_array = pad(img_array, padding=padding, fill=np.nan)
-    h_, w_, bands_ = img_array.shape
     mx = chunk_size * xres
     my = chunk_size * yres
-    X_points = np.arange(0, w_ - chunk_size + 1, step)
-    Y_points = np.arange(0, h_ - chunk_size + 1, step)
-    pred_img = np.empty((h_, w_), dtype=np.uint8)
+    padded = chunk_size * 2
+    h = input_image.height
+    w = input_image.width
+    h_ = h + padded
+    w_ = w + padded
+    dist_samples = int(round(chunk_size * (1 - 1.0 / 2.0)))
+
+    # switch to evaluate mode
+    model.eval()
+
+    # initialize test time augmentation
+    transforms = tta.Compose([tta.HorizontalFlip(), ])
+    # construct window for smoothing
+    WINDOW_SPLINE_2D = _window_2D(window_size=padded, power=2.0)
+    WINDOW_SPLINE_2D = torch.as_tensor(np.moveaxis(WINDOW_SPLINE_2D, 2, 0), ).type(torch.float)
+    WINDOW_SPLINE_2D = WINDOW_SPLINE_2D.to(device)
+
+    fp = np.memmap(tp_mem, dtype='float16', mode='w+', shape=(h_, w_, num_classes))
```

Comment:
can you add more comments from here on? Why do you create a memory map from tp_mem? What does it become further on?

Reply:
basically a temporary memory map file was introduced to aid inference on very large tifs